Applications/FileIO/GocadIO/GocadTSurfaceReader.cpp

```diff
@@ -0,0 +1,525 @@
+        pos = in.tellg();
+    }
+    ERR("%s", eof_error.c_str());
+    return false;
+}
+
+bool GocadTSurfaceReader::parseSurface(
+    std::ifstream& in,
+    std::vector<MeshLib::Node*>& nodes,
+    std::vector<MeshLib::Element*>& elems,
+    std::map<std::size_t, std::size_t>& node_id_map,
+    MeshLib::Properties& mesh_prop)
+{
+    if (!parseNodes(in, nodes, node_id_map, mesh_prop))
+        return false;
+    if (!parseElements(in, nodes, elems, node_id_map, mesh_prop))
+        return false;
+
+    std::string line("");
+    while (std::getline(in, line))
+    {
+        std::vector<std::string> str = BaseLib::splitString(line);
+        if (str[0] == "TFACE")
+        {
+            parseSurface(in, nodes, elems, node_id_map, mesh_prop);
+            return true;
+        }
+        else if (str[0] == "BSTONE")
+        {
+            // borderstone definition - currently ignored
+        }
+        else if (str[0] == "BORDER")
+        {
+            // border tracking direction - currently ignored
+        }
+        else if (line == "END")
+        {
+            return true;
+        }
+        else
+        {
+            WARN(
+                "GocadTSurfaceReader::parseSurface() - Unknown keyword found: "
+                "%s",
+                line.c_str());
+        }
+    }
+    ERR("%s", eof_error.c_str());
+    return false;
+}
+
+MeshLib::Node* createNode(std::stringstream& sstr)
+{
+    std::string keyword;
+    std::size_t id;
+    std::array<double, 3> data;
+    sstr >> keyword >> id >> data[0] >> data[1] >> data[2];
+    return new MeshLib::Node(data, id);
+}
+
+bool GocadTSurfaceReader::parseNodes(
+    std::ifstream& in,
+    std::vector<MeshLib::Node*>& nodes,
+    std::map<std::size_t, std::size_t>& node_id_map,
+    MeshLib::Properties& mesh_prop)
+{
+    NODE_TYPE t = NODE_TYPE::UNSPECIFIED;
+    double value;
+    std::vector<std::string> const array_names =
+        mesh_prop.getPropertyVectorNames();
+    std::streampos pos = in.tellg();
+    std::string line("");
+    while (std::getline(in, line))
+    {
+        std::vector<std::string> str = BaseLib::splitString(line);
+        if (line.substr(0, 4) == "TRGL")
+        {
+            in.seekg(pos);
+            return true;
+        }
+
+        if (line.empty() || isCommentLine(line))
+            continue;
+        if (!(line.substr(0, 4) == "VRTX" || line.substr(0, 5) == "PVRTX" ||
+              line.substr(0, 4) == "ATOM"))
+        {
+            WARN(
+                "GocadTSurfaceReader::parseNodes() - Unknown keyword found: %s",
+                line.c_str());
+            continue;
+        }
+
+        std::stringstream sstr(line);
+        if (line.substr(0, 4) == "VRTX" && t != NODE_TYPE::PVRTX)
+        {
+            nodes.push_back(createNode(sstr));
+        }
+        else if (line.substr(0, 5) == "PVRTX" && t != NODE_TYPE::VRTX)
+        {
+            nodes.push_back(createNode(sstr));
+            for (std::string name : array_names)
```

Comment:
Why access by value? Why not `for (std::string const& name : array_names)`?

Reply:
:heavy_check_mark: